File: src/client/java/io/github/romeoahmed/cursedoath/client/render/limitless/EnergyTrails.java
package io.github.romeoahmed.cursedoath.client.render.limitless;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import io.github.romeoahmed.cursedoath.client.render.EffectMesh;
import io.github.romeoahmed.cursedoath.technique.Technique;
import io.github.romeoahmed.cursedoath.technique.TechniqueTuning;
import net.minecraft.util.ARGB;
import net.minecraft.world.phys.Vec3;

/// Surface flows, soft glow, and directional trails around the opaque body.
public final class EnergyTrails {
    private static final double GLOW_PLANE = 1.01;
    private static final double GLOW_RADIUS = 1.5;
    private static final int GLOW_ALPHA = 100;
    private static final int GLOW_BANDS = 6;
    private static final int GLOW_SEGMENTS = 32;
    private static final int BLUE_RIBBONS = 5;
    private static final int RIBBON_STEPS = 10;
    private static final double BLUE_SPEED = 0.09;
    private static final double BLUE_SWEEP = 2.2;
    private static final double BLUE_REACH = 0.3;
    private static final double BLUE_TWIST = 0.18;
    private static final double SURFACE_OFFSET = 1.015;
    private static final double DARK_WIDTH = 0.18;
    private static final double LIGHT_WIDTH = 0.035;
    private static final int BLUE_INK = 0x062C85;
    private static final int BLUE_LIGHT = 0x24CCFF;
    private static final int RED_RAYS = 12;
    private static final double RED_SPEED = 0.11;
    private static final double RAY_START = 1.1;
    private static final double RAY_TRAVEL = 2.0;
    private static final double RAY_LENGTH = 1.8;
    private static final double RAY_WIDTH = 0.075;
    private static final int RED_LIGHT = 0xFF2948;
    private static final int PURPLE_FLOWS = 7;
    private static final double PURPLE_SPEED = 0.07;
    private static final double PURPLE_SWEEP = 2.0;
    private static final double FLOW_RIPPLE = 0.035;
    private static final int RIPPLE_FREQUENCY = 5;
    private static final double PURPLE_WIDTH = 0.04;
    private static final int PURPLE_LIGHT = 0xDC79FF;
    private static final float FLOW_ALPHA = 0.55f;
    private static final int WAKE_STREAMS = 12;
    private static final double WAKE_SPEED = 0.08;
    private static final double WAKE_RADIUS = 0.7;
    private static final double WAKE_START = 0.45;
    private static final double WAKE_LENGTH = 0.45;
    private static final double WAKE_SPREAD = 0.12;
    private static final double WAKE_WIDTH = 0.035;
    private static final double IMPACT_WIDTH = 0.3;
    private static final double IMPACT_INNER = 0.6;
    private static final double GOLDEN_ANGLE = 2.399963;
    private static final int NORMAL_COUNT = 24;
    private static final Vec3[] NORMALS = new Vec3[NORMAL_COUNT], CIRCLE = new Vec3[GLOW_SEGMENTS + 1];
    private static final EffectMesh.Basis[] AXES = new EffectMesh.Basis[NORMAL_COUNT];

    static {
        for (int i = 0; i < NORMAL_COUNT; i++) {
            double y = 1 - 2 * (i + 0.5) / NORMAL_COUNT, radius = Math.sqrt(1 - y * y);
            NORMALS[i] = new Vec3(Math.cos(i * GOLDEN_ANGLE) * radius, y, Math.sin(i * GOLDEN_ANGLE) * radius);
            AXES[i] = EffectMesh.basis(NORMALS[i]);
        }
        for (int i = 0; i <= GLOW_SEGMENTS; i++) {
            double angle = i * 2 * Math.PI / GLOW_SEGMENTS;
            CIRCLE[i] = new Vec3(Math.cos(angle), Math.sin(angle), 0);
        }
    }

    private final PoseStack.Pose pose;
    private final VertexConsumer vertices;
    private final EffectMesh mesh;

    public EnergyTrails(PoseStack.Pose pose, VertexConsumer vertices) {
        this.pose = pose;
        this.vertices = vertices;
        mesh = new EffectMesh(pose, vertices, 1, 1, 255);
    }

    public void draw(LimitlessEffects.Form form) {
        glow(form);
        switch (form.technique()) {
            case BLUE -> blueRibbons(form, false);
            case RED -> redRays(form);
            case PURPLE -> {
                purpleFlow(form);
                if (form.flying()) wake(form);
            }
            default -> {}
        }
    }

    private void glow(LimitlessEffects.Form form) {
        var view = EnergySurface.viewFrom(pose, form.center()).normalize();
        var axes = EffectMesh.basis(view);
        // Keep the billboard in front of the body; intersection cuts a dark disc into the glow.
        var center = form.center().add(view.scale(form.radius() * GLOW_PLANE));
        int color = ARGB.srgbLerp(form.fusion(), form.technique().color(), Technique.PURPLE.color());
        float strength = 1 - form.fusion() / 2;
        for (int band = 0; band < GLOW_BANDS; band++) {
            double inner = (double) band / GLOW_BANDS,
                    outer = (band + 1.0) / GLOW_BANDS,
                    innerRadius = form.radius() * GLOW_RADIUS * inner,
                    outerRadius = form.radius() * GLOW_RADIUS * outer;
            int innerColor = ARGB.color((int) ((1 - inner) * (1 - inner) * GLOW_ALPHA * strength), color),
                    outerColor = ARGB.color((int) ((1 - outer) * (1 - outer) * GLOW_ALPHA * strength), color);
            for (int i = 0; i < GLOW_SEGMENTS; i++) {
                glowVertex(center, axes, CIRCLE[i], innerRadius, innerColor);
                glowVertex(center, axes, CIRCLE[i + 1], innerRadius, innerColor);
                glowVertex(center, axes, CIRCLE[i + 1], outerRadius, outerColor);
                glowVertex(center, axes, CIRCLE[i], outerRadius, outerColor);
            }
        }
    }

    private void glowVertex(Vec3 center, EffectMesh.Basis axes, Vec3 point, double radius, int color) {
        double x = point.x * radius, y = point.y * radius;
        var right = axes.side();
        var up = axes.up();
        vertices.addVertex(
                        pose,
                        (float) (center.x + right.x * x + up.x * y),
                        (float) (center.y + right.y * x + up.y * y),
                        (float) (center.z + right.z * x + up.z * y))
                .setColor(color);
    }

    public void blueRibbons(LimitlessEffects.Form form, boolean dark) {
        for (int i = 0; i < BLUE_RIBBONS; i++) {
            int index = i * NORMAL_COUNT / BLUE_RIBBONS;
            var axis = NORMALS[index];
            var axes = AXES[index];
            double phase = i * GOLDEN_ANGLE - form.age() * BLUE_SPEED;
            for (int step = 0; step < RIBBON_STEPS; step++) {
                double t = (double) step / RIBBON_STEPS;
                var width = axis.scale(form.radius()
                        * (dark ? DARK_WIDTH : LIGHT_WIDTH)
                        * Math.sin(t * Math.PI)
                        * (1 - form.fusion()));
                var a = bluePoint(form, axis, axes, phase, t);
                var b = bluePoint(form, axis, axes, phase, (step + 1.0) / RIBBON_STEPS);
                mesh.ribbon(a, b, width, dark ? BLUE_INK : BLUE_LIGHT);
            }
        }
    }

    private static Vec3 bluePoint(
            LimitlessEffects.Form form, Vec3 axis, EffectMesh.Basis axes, double phase, double t) {
        double angle = phase + t * BLUE_SWEEP,
                radius = form.radius() * (SURFACE_OFFSET + BLUE_REACH * (1 - t) * (1 - t));
        return form.center()
                .add(axes.side().scale(Math.cos(angle) * radius))
                .add(axes.up().scale(Math.sin(angle) * radius))
                .add(axis.scale(form.radius() * BLUE_TWIST * Math.sin(t * Math.PI)));
    }

    private void redRays(LimitlessEffects.Form form) {
        for (int i = 0; i < RED_RAYS; i++) {
            int index = i * NORMAL_COUNT / RED_RAYS;
            var direction = NORMALS[index];
            double phase = (form.age() * RED_SPEED + (double) i / RED_RAYS) % 1,
                    reach = form.radius() * (RAY_START + phase * RAY_TRAVEL);
            var width = AXES[index].side().scale(form.radius() * RAY_WIDTH * (1 - phase) * (1 - form.fusion()));
            var a = form.center().add(direction.scale(reach));
            var b = form.center().add(direction.scale(reach + form.radius() * RAY_LENGTH));
            mesh.slash(a, b, width, RED_LIGHT, (float) (1 - phase));
        }
    }

    private void purpleFlow(LimitlessEffects.Form form) {
        for (int i = 0; i < PURPLE_FLOWS; i++) {
            int index = i * NORMAL_COUNT / PURPLE_FLOWS;
            var axis = NORMALS[index];
            var right = AXES[index].side();
            var up = AXES[index].up();
            double phase = i * GOLDEN_ANGLE + form.age() * PURPLE_SPEED;
            for (int step = 0; step < RIBBON_STEPS; step++) {
                double t = (double) step / RIBBON_STEPS,
                        angle = phase + t * PURPLE_SWEEP,
                        radius =
                                form.radius()
                                        * (SURFACE_OFFSET + FLOW_RIPPLE * (1 + Math.sin(angle * RIPPLE_FREQUENCY)));
                var a = form.center()
                        .add(right.scale(Math.cos(angle) * radius))
                        .add(up.scale(Math.sin(angle) * radius));
                double next = angle + PURPLE_SWEEP / RIBBON_STEPS;
                var b = form.center().add(right.scale(Math.cos(next) * radius)).add(up.scale(Math.sin(next) * radius));
                var width = axis.scale(form.radius() * PURPLE_WIDTH * Math.sin(t * Math.PI));
                mesh.ribbon(
                        a,
                        b,
                        width,
                        i % 2 == 0 ? PURPLE_LIGHT : form.technique().color(),
                        FLOW_ALPHA);
            }
        }
    }

    private void wake(LimitlessEffects.Form form) {
        var axes = EffectMesh.basis(form.direction());
        var right = axes.side();
        var up = axes.up();
        for (int i = 0; i < WAKE_STREAMS; i++) {
            double phase = (form.age() * WAKE_SPEED + (double) i / WAKE_STREAMS) % 1, angle = i * GOLDEN_ANGLE;
            var radial = right.scale(Math.cos(angle)).add(up.scale(Math.sin(angle)));
            var start = form.center()
                    .add(radial.scale(form.radius() * WAKE_RADIUS))
                    .subtract(form.direction().scale(form.radius() * (WAKE_START + phase)));
            var end = start.subtract(form.direction().scale(form.radius() * WAKE_LENGTH))
                    .add(radial.scale(form.radius() * WAKE_SPREAD));
            mesh.slash(start, end, radial.scale(form.radius() * WAKE_WIDTH), PURPLE_LIGHT, (float)
                    ((1 - phase) * FLOW_ALPHA));
        }
    }

    public void impact(float progress) {
        double radius = Math.sqrt(progress) * TechniqueTuning.RED_RADIUS;
        float alpha = (1 - progress) * (1 - progress);
        for (int index = 0; index < NORMAL_COUNT; index++) {
            var normal = NORMALS[index];
            var width = AXES[index].side().scale(IMPACT_WIDTH * alpha);
            mesh.slash(normal.scale(radius * IMPACT_INNER), normal.scale(radius), width, RED_LIGHT, alpha);
        }
    }
}

File: src/client/kotlin/io/github/romeoahmed/cursedoath/client/render/limitless/EnergyTrails.kt
package io.github.romeoahmed.cursedoath.client.render.limitless

import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.VertexConsumer
import io.github.romeoahmed.cursedoath.client.render.EffectMesh
import io.github.romeoahmed.cursedoath.technique.Technique
import io.github.romeoahmed.cursedoath.technique.TechniqueTuning
import net.minecraft.util.ARGB
import net.minecraft.world.phys.Vec3
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/** Surface flows, soft glow, and directional trails around the opaque body. */
internal class EnergyTrails(
    private val pose: PoseStack.Pose,
    private val vertices: VertexConsumer,
    private val opacity: Float = 1f,
) {
    private val mesh = EffectMesh(pose, vertices, opacity, maxAlpha = 255)

    fun draw(form: LimitlessEffects.Form) {
        glow(form)
        when (form.technique) {
            Technique.BLUE -> {
                blueRibbons(form)
            }

            Technique.RED -> {
                redRays(form)
            }

            Technique.PURPLE -> {
                purpleFlow(form)
                if (form.flying) wake(form)
            }

            else -> {}
        }
    }

    private fun glow(form: LimitlessEffects.Form) {
        val view = EnergySurface.viewFrom(pose, form.center).normalize()
        val (right, up) = EffectMesh.basis(view)
        // Keep the billboard in front of the whole body; intersecting it cuts a dark disc into the glow.
        val center = form.center.add(view.scale(form.radius * GLOW_PLANE))
        val color = ARGB.srgbLerp(form.fusion, form.technique.color, Technique.PURPLE.color)
        val strength = 1 - form.fusion / 2

        fun vertex(
            point: Vec3,
            radius: Double,
            color: Int,
        ) {
            val x = point.x * radius
            val y = point.y * radius
            vertices
                .addVertex(
                    pose,
                    (center.x + right.x * x + up.x * y).toFloat(),
                    (center.y + right.y * x + up.y * y).toFloat(),
                    (center.z + right.z * x + up.z * y).toFloat(),
                ).setColor(color)
        }
        for (band in 0..<GLOW_BANDS) {
            val inner = band.toDouble() / GLOW_BANDS
            val outer = (band + 1.0) / GLOW_BANDS
            val innerRadius = form.radius * GLOW_RADIUS * inner
            val outerRadius = form.radius * GLOW_RADIUS * outer
            val innerColor = ARGB.color(((1 - inner) * (1 - inner) * opacity * GLOW_ALPHA * strength).toInt(), color)
            val outerColor = ARGB.color(((1 - outer) * (1 - outer) * opacity * GLOW_ALPHA * strength).toInt(), color)
            for (i in 0..<GLOW_SEGMENTS) {
                vertex(circle[i], innerRadius, innerColor)
                vertex(circle[i + 1], innerRadius, innerColor)
                vertex(circle[i + 1], outerRadius, outerColor)
                vertex(circle[i], outerRadius, outerColor)
            }
        }
    }

    fun blueRibbons(
        form: LimitlessEffects.Form,
        dark: Boolean = false,
    ) {
        for (i in 0..<BLUE_RIBBONS) {
            val seed = i * GOLDEN_ANGLE
            val index = i * NORMAL_COUNT / BLUE_RIBBONS
            val axis = normals[index]
            val (right, up) = axes[index]
            val phase = seed - form.age * BLUE_SPEED

            fun point(t: Double): Vec3 {
                val angle = phase + t * BLUE_SWEEP
                val radius = form.radius * (SURFACE_OFFSET + BLUE_REACH * (1 - t) * (1 - t))
                return form.center
                    .add(right.scale(cos(angle) * radius))
                    .add(up.scale(sin(angle) * radius))
                    .add(axis.scale(form.radius * BLUE_TWIST * sin(t * PI)))
            }
            for (step in 0..<RIBBON_STEPS) {
                val t = step.toDouble() / RIBBON_STEPS
                val width =
                    axis.scale(form.radius * (if (dark) DARK_WIDTH else LIGHT_WIDTH) * sin(t * PI) * (1 - form.fusion))
                val a = point(t)
                val b = point((step + 1.0) / RIBBON_STEPS)
                mesh.ribbon(a, b, width, if (dark) BLUE_INK else BLUE_LIGHT)
            }
        }
    }

    private fun redRays(form: LimitlessEffects.Form) {
        for (i in 0..<RED_RAYS) {
            val index = i * NORMAL_COUNT / RED_RAYS
            val direction = normals[index]
            val phase = (form.age * RED_SPEED + i.toDouble() / RED_RAYS) % 1.0
            val reach = form.radius * (RAY_START + phase * RAY_TRAVEL)
            val width =
                axes[index].first.scale(form.radius * RAY_WIDTH * (1 - phase) * (1 - form.fusion))
            val a = form.center.add(direction.scale(reach))
            val b = form.center.add(direction.scale(reach + form.radius * RAY_LENGTH))
            mesh.slash(a, b, width, RED_LIGHT, (1 - phase).toFloat())
        }
    }

    private fun purpleFlow(form: LimitlessEffects.Form) {
        for (i in 0..<PURPLE_FLOWS) {
            val index = i * NORMAL_COUNT / PURPLE_FLOWS
            val axis = normals[index]
            val (right, up) = axes[index]
            val phase = i * GOLDEN_ANGLE + form.age * PURPLE_SPEED
            for (step in 0..<RIBBON_STEPS) {
                val t = step.toDouble() / RIBBON_STEPS
                val angle = phase + t * PURPLE_SWEEP
                val radius = form.radius * (SURFACE_OFFSET + FLOW_RIPPLE * (1 + sin(angle * RIPPLE_FREQUENCY)))
                val a = form.center.add(right.scale(cos(angle) * radius)).add(up.scale(sin(angle) * radius))
                val next = angle + PURPLE_SWEEP / RIBBON_STEPS
                val b = form.center.add(right.scale(cos(next) * radius)).add(up.scale(sin(next) * radius))
                val width = axis.scale(form.radius * PURPLE_WIDTH * sin(t * PI))
                mesh.ribbon(a, b, width, if (i % 2 == 0) PURPLE_LIGHT else form.technique.color, FLOW_ALPHA)
            }
        }
    }

    private fun wake(form: LimitlessEffects.Form) {
        val (right, up) = EffectMesh.basis(form.direction)
        for (i in 0..<WAKE_STREAMS) {
            val phase = (form.age * WAKE_SPEED + i.toDouble() / WAKE_STREAMS) % 1.0
            val angle = i * GOLDEN_ANGLE
            val radial = right.scale(cos(angle)).add(up.scale(sin(angle)))
            val start =
                form.center
                    .add(radial.scale(form.radius * WAKE_RADIUS))
                    .subtract(form.direction.scale(form.radius * (WAKE_START + phase)))
            val end =
                start
                    .subtract(form.direction.scale(form.radius * WAKE_LENGTH))
                    .add(radial.scale(form.radius * WAKE_SPREAD))
            mesh.slash(
                start,
                end,
                radial.scale(form.radius * WAKE_WIDTH),
                PURPLE_LIGHT,
                ((1 - phase) * FLOW_ALPHA).toFloat(),
            )
        }
    }

    fun impact(progress: Float) {
        val radius = sqrt(progress.toDouble()) * TechniqueTuning.RED_RADIUS
        val alpha = (1 - progress) * (1 - progress)
        for ((index, normal) in normals.withIndex()) {
            val width = axes[index].first.scale(IMPACT_WIDTH * alpha)
            mesh.slash(normal.scale(radius * IMPACT_INNER), normal.scale(radius), width, RED_LIGHT, alpha)
        }
    }

    private companion object {
        const val GLOW_PLANE = 1.01
        const val GLOW_RADIUS = 1.5
        const val GLOW_ALPHA = 100
        const val GLOW_BANDS = 6
        const val GLOW_SEGMENTS = 32
        const val BLUE_RIBBONS = 5
        const val RIBBON_STEPS = 10
        const val BLUE_SPEED = 0.09
        const val BLUE_SWEEP = 2.2
        const val BLUE_REACH = 0.3
        const val BLUE_TWIST = 0.18
        const val SURFACE_OFFSET = 1.015
        const val DARK_WIDTH = 0.18
        const val LIGHT_WIDTH = 0.035
        const val BLUE_INK = 0x062C85
        const val BLUE_LIGHT = 0x24CCFF
        const val RED_RAYS = 12
        const val RED_SPEED = 0.11
        const val RAY_START = 1.1
        const val RAY_TRAVEL = 2.0
        const val RAY_LENGTH = 1.8
        const val RAY_WIDTH = 0.075
        const val RED_LIGHT = 0xFF2948
        const val PURPLE_FLOWS = 7
        const val PURPLE_SPEED = 0.07
        const val PURPLE_SWEEP = 2.0
        const val FLOW_RIPPLE = 0.035
        const val RIPPLE_FREQUENCY = 5
        const val PURPLE_WIDTH = 0.04
        const val PURPLE_LIGHT = 0xDC79FF
        const val FLOW_ALPHA = 0.55f
        const val WAKE_STREAMS = 12
        const val WAKE_SPEED = 0.08
        const val WAKE_RADIUS = 0.7
        const val WAKE_START = 0.45
        const val WAKE_LENGTH = 0.45
        const val WAKE_SPREAD = 0.12
        const val WAKE_WIDTH = 0.035
        const val IMPACT_WIDTH = 0.3
        const val IMPACT_INNER = 0.6
        const val GOLDEN_ANGLE = 2.399963
        const val NORMAL_COUNT = 24
        val normals =
            Array(NORMAL_COUNT) { i ->
                val y = 1 - 2 * (i + 0.5) / NORMAL_COUNT
                val radius = sqrt(1 - y * y)
                Vec3(cos(i * GOLDEN_ANGLE) * radius, y, sin(i * GOLDEN_ANGLE) * radius)
            }
        val axes = Array(NORMAL_COUNT) { EffectMesh.basis(normals[it]) }
        val circle =
            Array(GLOW_SEGMENTS + 1) { i ->
                val angle = i * 2 * PI / GLOW_SEGMENTS
                Vec3(cos(angle), sin(angle), 0.0)
            }
    }
}
